### PR TITLE
Add separate ops for collecting metrics from multiple namespaces in scheduler_perf

### PR DESCRIPTION
**Pull request by macsko**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This PR adds a possibility to explicitly define when to start and stop collecting metrics within scheduler_perf test case. Especially, it allows to collect metrics from multiple namespaces, so to include many createPods operations. It is needed for queueing hints benchmark test cases. 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/integration/scheduler_perf/scheduler_perf.go [hunk 1]
```diff
@@ -779,6 +783,58 @@ func (so sleepOp) patchParams(_ *workload) (realOp, error) {
 	return &so, nil
 }
 
+// startCollectingMetricsOp defines an op that starts metrics collectors.
+// stopCollectingMetricsOp has to be used after this op to finish collecting.
+type startCollectingMetricsOp struct {
+	// Must be "startCollectingMetrics".
+	Opcode operationCode
+	// Name appended to workload's name.
+	Name string
```
Comment (sanposhiho) [inline test/integration/scheduler_perf/scheduler_perf.go:792]: what is it for?

File: test/integration/scheduler_perf/scheduler_perf.go [hunk 2]
```diff
@@ -1116,6 +1172,52 @@ func runWorkload(tCtx ktesting.TContext, tc *testCase, w *workload, informerFact
...
+			}()
+		}
+		return func() {
+			collectorCtx.Cancel("cleaning up")
+		}
+	}
+	stopCollectingMetrics := func(opIndex int) {
+		collectorCtx.Cancel("collecting metrics, collector must stop first")
+		collectorWG.Wait()
+		mu.Lock()
```
[the remark this answers is not in the document]
Comment (macsko) [reply test/integration/scheduler_perf/scheduler_perf.go:1209]: Right, we don't need it

File: test/integration/scheduler_perf/scheduler_perf.go [hunk 3]
```diff
@@ -1116,6 +1172,52 @@ func runWorkload(tCtx ktesting.TContext, tc *testCase, w *workload, informerFact
...
+		for _, collector := range collectors {
+			items := collector.collect()
+			dataItems = append(dataItems, items...)
+			err := compareMetricWithThreshold(items, w.Threshold, *w.ThresholdMetricSelector)
+			if err != nil {
+				tCtx.Errorf("op %d: %s", opIndex, err)
+			}
+		}
+		mu.Unlock()
+	}
```
[the remark this answers is not in the document]
Comment (macsko) [reply test/integration/scheduler_perf/scheduler_perf.go]: They just share many arguments with the workload. I moved them out of `runWorkload` for clearer dependencies.

File: test/integration/scheduler_perf/scheduler_perf.go (hunk 1 above)
Comment (macsko) [reply test/integration/scheduler_perf/scheduler_perf.go:828]: For default metrics collection within createPodsOp we append namespace name to the name of workload in the results json. In this case I added this argument to use some user-defined name if we can't use the namespace directly. Especially we need to remember that collecting the metrics multiple time within one workload is not forbidden as long as they don't interleave.

File: test/integration/scheduler_perf/scheduler_perf.go [hunk 4]
```diff
@@ -1440,6 +1506,15 @@ func runWorkload(tCtx ktesting.TContext, tc *testCase, w *workload, informerFact
 			case <-tCtx.Done():
 			case <-time.After(concreteOp.Duration):
 			}
+
+		case *startCollectingMetricsOp:
```
Comment (sanposhiho) [inline test/integration/scheduler_perf/scheduler_perf.go:1510]: To catch a mistake of overwrapping.

```go
		case *startCollectingMetricsOp:
			if collectorCtx != nil { error }

....

		case *stopCollectingMetricsOp:
			items := stopCollectingMetrics(tCtx, collectorCtx, &collectorWG, w.Threshold, *w.ThresholdMetricSelector, opIndex, collectors)
			dataItems = append(dataItems, items...)
			collectorCtx = nil
```

File: test/integration/scheduler_perf/scheduler_perf.go [hunk 5]
```diff
@@ -1110,6 +1166,43 @@ func checkEmptyInFlightEvents() error {
...
+		go func() {
+			defer collectorWG.Done()
+			collector.run(collectorCtx)
+		}()
+	}
+	return collectorCtx, collectors
+}
+
+func stopCollectingMetrics(tCtx ktesting.TContext, collectorCtx ktesting.TContext, collectorWG *sync.WaitGroup, threshold float64, tms thresholdMetricSelector, opIndex int, collectors []testDataCollector) []DataItem {
+	collectorCtx.Cancel("collecting metrics, collector must stop first")
```
Comment (sanposhiho) [inline test/integration/scheduler_perf/scheduler_perf.go:1192]: To catch a mistake of missing start.

```suggestion
	if collectorCtx == nil { error }
	
	collectorCtx.Cancel("collecting metrics, collector must stop first")
```

**Review by sanposhiho — COMMENTED** [review]
Two small nits

File: test/integration/scheduler_perf/scheduler_perf.go (hunk 4 above)
Comment (macsko) [reply test/integration/scheduler_perf/scheduler_perf.go:1517]: Good ideas, done

**Review by sanposhiho — COMMENTED** [review]
/lgtm

File: test/integration/scheduler_perf/scheduler_perf.go [hunk 6]
```diff
@@ -815,6 +819,58 @@ func (so sleepOp) patchParams(_ *workload) (realOp, error) {
 	return &so, nil
 }
 
+// startCollectingMetricsOp defines an op that starts metrics collectors.
+// stopCollectingMetricsOp has to be used after this op to finish collecting.
+type startCollectingMetricsOp struct {
+	// Must be "startCollectingMetrics".
+	Opcode operationCode
+	// Name appended to workload's name.
```
Comment (alculquicondor) [inline test/integration/scheduler_perf/scheduler_perf.go:827]:
```suggestion
	// Name appended to workload's name in results.
```

**Review by alculquicondor — COMMENTED** [review]
/approve
/hold for nit

File: test/integration/scheduler_perf/scheduler_perf.go (hunk 6 above)
Comment (macsko) [reply test/integration/scheduler_perf/scheduler_perf.go]: Done

**Pull request merged**